File: 1. Longest Common Subsequence.cpp
// Task: Given two strings, text1 and text2, find and return their longest common subsequence as a
// new string. If there are multiple solutions, any one will suffice.

#include <iostream>
#include <vector>
#include <string>
using namespace std;

// TC: O(m * n) where m and n are lengths of text1 and text2
// SC: O(m * n) for the DP table
string longestCommonSubsequence(string text1, string text2)
{
    int m = text1.length();
    int n = text2.length();

    // Create DP table
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));

    // Fill the DP table
    for (int i = 1; i <= m; i++)
    {
        for (int j = 1; j <= n; j++)
        {
            if (text1[i - 1] == text2[j - 1])
            {
                dp[i][j] = dp[i - 1][j - 1] + 1;
            }
            else
            {
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
            }
        }
    }

    // Backtrack to find the LCS string
    string lcs = "";
    int i = m, j = n;
    while (i > 0 && j > 0)
    {
        if (text1[i - 1] == text2[j - 1])
        {
            lcs = text1[i - 1] + lcs;
            i--;
            j--;
        }
        else if (dp[i - 1][j] > dp[i][j - 1])
        {
            i--;
        }
        else
        {
            j--;
        }
    }

    return lcs;
}

int main()
{
    string text1, text2;
    cout << "Enter first string: ";
    cin >> text1;
    cout << "Enter second string: ";
    cin >> text2;

    string result = longestCommonSubsequence(text1, text2);
    cout << "Longest Common Subsequence: " << result << endl;
    cout << "Length: " << result.length() << endl;

    return 0;
}

File: 1. Longest Common Subsequence.py
# Task: Given two strings, text1 and text2, find and return their longest common subsequence as a new string. If there are multiple solutions, any one will suffice.

# TC: O(m * n) where m and n are lengths of text1 and text2
# SC: O(m * n) for the DP table
def longest_common_subsequence(text1, text2):
    m, n = len(text1), len(text2)
    
    # Create DP table
    dp = [[0] * (n + 1) for _ in range(m + 1)]
    
    # Fill the DP table
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if text1[i - 1] == text2[j - 1]:
                dp[i][j] = dp[i - 1][j - 1] + 1
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])
    
    # Backtrack to find the LCS string
    lcs = []
    i, j = m, n
    while i > 0 and j > 0:
        if text1[i - 1] == text2[j - 1]:
            lcs.append(text1[i - 1])
            i -= 1
            j -= 1
        elif dp[i - 1][j] > dp[i][j - 1]:
            i -= 1
        else:
            j -= 1
    
    return ''.join(reversed(lcs))

if __name__ == "__main__":
    text1 = input("Enter first string: ")
    text2 = input("Enter second string: ")
    
    result = longest_common_subsequence(text1, text2)
    print(f"Longest Common Subsequence: {result}")
    print(f"Length: {len(result)}")
